src/vanilla/utils/proxyMap.ts

```diff
@@ -1,111 +1,153 @@
-import { proxy } from '../../vanilla.ts'
+import { getVersion, proxy } from '../../vanilla.ts'
 
-type KeyValRecord<K, V> = [key: K, value: V]
+const maybeProxify = (x: any) => proxy({ x }).x
 
-type InternalProxyMap<K, V> = Map<K, V> & {
-  data: KeyValRecord<K, V>[]
-  toJSON: object
+const isProxy = (x: any) => getVersion(x) !== undefined
+
+type InternalProxyObject<K, V> = Map<K, V> & {
+  data: Array<[K, V | undefined]>
+  size: number
+  toJSON: () => Map<K, V>
 }
 
-/**
- * proxyMap
- *
- * This is to create a proxy which mimic the native Map behavior.
- * The API is the same as Map API
- *
- * @example
- * import { proxyMap } from 'valtio/utils'
- * const state = proxyMap([["key", "value"]])
- *
- * //can be used inside a proxy as well
- * const state = proxy({
- *   count: 1,
- *   map: proxyMap()
- * })
- *
- * // When using an object as a key, you can wrap it with `ref` so it's not proxied
- * // this is useful if you want to preserve the key equality
- * import { ref } from 'valtio'
- *
- * const key = ref({})
- * state.set(key, "value")
- * state.get(key) //value
- *
- * const key = {}
- * state.set(key, "value")
- * state.get(key) //undefined
- */
-export function proxyMap<K, V>(entries?: Iterable<readonly [K, V]> | null) {
-  const map: InternalProxyMap<K, V> = proxy({
-    data: Array.from(entries || []) as KeyValRecord<K, V>[],
-    has(key) {
-      return this.data.some((p) => p[0] === key)
+export function proxyMap<K, V>(entries?: Iterable<[K, V]> | undefined | null) {
+  const data: Array<[K, V]> = []
+  const indexMap = new Map<K, number>()
+  const emptyIndexes: number[] = []
+
+  if (entries !== null && typeof entries !== 'undefined') {
+    if (typeof entries[Symbol.iterator] !== 'function') {
+      throw new TypeError(
+        'proxyMap:\n\tinitial state must be iterable\n\t\ttip: structure should be [[key, value]]',
+      )
+    }
+    for (const [k, v] of entries) {
+      const key = maybeProxify(k)
+      const value = maybeProxify(v)
+      indexMap.set(key, data.length)
+      data.push([key, value])
+    }
+  }
+
+  const vObject: InternalProxyObject<K, V> = {
+    data,
+    get size() {
+      return indexMap.size
+    },
+    get(key: K) {
+      const k = maybeProxify(key)
+      if (indexMap.has(k)) {
+        const index = indexMap.get(k)
+        if (this.data[index!] !== undefined) {
```

Comment:
If it's undefined, it means inconsistent `this.data`, right? Maybe we should throw an error?

Reply:
There was a reason I was doing this at some point...some edge case I can't think of at the moment. Let me get back to you on this.